File: backend/api/routes/notifications.py
"""
Notification API routes
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from api_database import get_db
from api_schemas.notification import (
    NotificationCreate,
    NotificationUpdate,
    NotificationResponse,
    NotificationListResponse
)
from api_services.notification_service import NotificationService
from typing import Optional

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListResponse)
async def get_notifications(
    user_role: Optional[str] = Query(None, description="Filter by user role (admin/staff)"),
    staff_id: Optional[int] = Query(None, description="Filter by staff ID"),
    is_read: Optional[bool] = Query(None, description="Filter by read status"),
    notification_type: Optional[str] = Query(None, description="Filter by notification type"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: AsyncSession = Depends(get_db)
):
    """Get notifications with optional filters"""
    service = NotificationService(db)
    
    notifications, total, unread_count = await service.get_notifications(
        user_role=user_role,
        staff_id=staff_id,
        is_read=is_read,
        notification_type=notification_type,
        skip=skip,
        limit=limit
    )
    
    return NotificationListResponse(
        notifications=[NotificationResponse.from_orm(n) for n in notifications],
        total=total,
        unread_count=unread_count
    )


@router.post("", response_model=NotificationResponse, status_code=201)
async def create_notification(
    notification: NotificationCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create a new notification (admin use)"""
    service = NotificationService(db)
    created = await service.create_notification(notification)
    return NotificationResponse.from_orm(created)


@router.patch("/{notification_id}", response_model=NotificationResponse)
async def update_notification(
    notification_id: int,
    update_data: NotificationUpdate,
    db: AsyncSession = Depends(get_db)
):
    """Update notification (mark as read/unread)"""
    service = NotificationService(db)
    
    if update_data.is_read:
        notification = await service.mark_as_read(notification_id)
    else:
        raise HTTPException(status_code=400, detail="Only marking as read is supported")
    
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    
    return NotificationResponse.from_orm(notification)


@router.post("/mark-all-read")
async def mark_all_as_read(
    user_role: Optional[str] = Query(None),
    staff_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """Mark all notifications as read for a user/role"""
    service = NotificationService(db)
    count = await service.mark_all_as_read(user_role=user_role, staff_id=staff_id)
    return {"marked_read": count}


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: int,
    db: AsyncSession = Depends(get_db)
):
    """Delete a notification"""
    service = NotificationService(db)
    success = await service.delete_notification(notification_id)
    
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    
    return {"message": "Notification deleted successfully"}


@router.get("/unread-count")
async def get_unread_count(
    user_role: Optional[str] = Query(None),
    staff_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """Get count of unread notifications"""
    service = NotificationService(db)
    _, _, unread_count = await service.get_notifications(
        user_role=user_role,
        staff_id=staff_id,
        is_read=False,
        limit=0
    )
    
    return {"unread_count": unread_count}

File: frontend/src/lib/api/notifications.ts
/**
 * Notification API client
 */
import { apiClient } from './client';

export interface Notification {
  id: number;
  user_role: string | null;
  staff_id: number | null;
  type: string;
  title: string;
  message: string;
  data: Record<string, any> | null;
  is_read: boolean;
  created_at: string;
  read_at: string | null;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
}

export interface NotificationFilters {
  user_role?: string;
  staff_id?: number;
  is_read?: boolean;
  notification_type?: string;
  skip?: number;
  limit?: number;
}

export const notificationApi = {
  /**
   * Get notifications with filters
   */
  getNotifications: async (filters?: NotificationFilters): Promise<NotificationListResponse> => {
    const params = new URLSearchParams();
    
    if (filters?.user_role) params.append('user_role', filters.user_role);
    if (filters?.staff_id) params.append('staff_id', filters.staff_id.toString());
    if (filters?.is_read !== undefined) params.append('is_read', filters.is_read.toString());
    if (filters?.notification_type) params.append('notification_type', filters.notification_type);
    if (filters?.skip) params.append('skip', filters.skip.toString());
    if (filters?.limit) params.append('limit', filters.limit.toString());
    
    const response = await apiClient.get<NotificationListResponse>(
      `/notifications?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Mark a notification as read
   */
  markAsRead: async (notificationId: number): Promise<Notification> => {
    const response = await apiClient.patch<Notification>(
      `/notifications/${notificationId}`,
      { is_read: true }
    );
    return response.data;
  },

  /**
   * Mark all notifications as read
   */
  markAllAsRead: async (filters?: { user_role?: string; staff_id?: number }): Promise<{ marked_read: number }> => {
    const params = new URLSearchParams();
    if (filters?.user_role) params.append('user_role', filters.user_role);
    if (filters?.staff_id) params.append('staff_id', filters.staff_id.toString());
    
    const response = await apiClient.post<{ marked_read: number }>(
      `/notifications/mark-all-read?${params.toString()}`
    );
    return response.data;
  },

  /**
   * Delete a notification
   */
  deleteNotification: async (notificationId: number): Promise<void> => {
    await apiClient.delete(`/notifications/${notificationId}`);
  },

  /**
   * Get unread notification count
   */
  getUnreadCount: async (filters?: { user_role?: string; staff_id?: number }): Promise<number> => {
    const params = new URLSearchParams();
    if (filters?.user_role) params.append('user_role', filters.user_role);
    if (filters?.staff_id) params.append('staff_id', filters.staff_id.toString());
    
    const response = await apiClient.get<{ unread_count: number }>(
      `/notifications/unread-count?${params.toString()}`
    );
    return response.data.unread_count;
  },
};
